### ISSUE TITLE
Add "change log" to update card

### ISSUE THREAD

**Issue by simurai** [atom/settings-view#282]
When updating packages, it would be nice to know what changed in the new version. Here a possible mockup that is based on the #277 branch:

![update](https://cloud.githubusercontent.com/assets/378023/5390910/ac8da044-8156-11e4-8624-5ec97ce6055b.png)

**Comment by Stanzilla** [comment]
Looks very good, and I agree that the "more" links should stand out a little more.

**Comment by segphault** [comment]
:+1: Very nice. I'd really like to see something like this implemented.

**Comment by braver** [comment]
I was just thinking of this: there is no method of communicating important changes to package users. Something like this would definitely help. :+1:

**Comment by abe33** [comment]
:+1: 
And like @mnquintana I think the `...more` link could be changed a bit, maybe wrapped to its own line below the excerpt.

**Comment by Stanzilla** [comment]
Also maybe hide the changelog behind a button/link you have to press first, could take a log of space when there are more updates with longer changelogs.

**Comment by mnquintana** [comment]
@Stanzilla I don't think that would be necessary because the changelog would be truncated in the update cards (truncated to 3 lines here)

**Comment by Stanzilla** [comment]
Oh indeed, that makes sense.

**Comment by thedaniel** [comment]
I'd rather do something this than truncate the changelog, as long as it doesn't break the columnar layout:

![expando](https://cloud.githubusercontent.com/assets/1476/6771789/80a22bf8-d0a7-11e4-8313-ac32e9c2d81a.gif)

**Comment by davidspiess** [comment]
The mockup looks great! Hopefully this gets implemented :)

**Comment by zacharytamas** [comment]
I'd like to provide this to users of my packages as well.

**Comment by piohhmy** [comment]
This would be great!

**Comment by TrevorJTClarke** [comment]
+1! This would be awesome!

**Comment by zenopopovici** [comment]
I was just about to open an issue regarding this. It's really important to see what's changed.

**Comment by jerone** [comment]
:+1:

**Comment by steelbrain** [comment]
Here's my :+1: on that

**Comment by eboracus** [comment]
This would be fantastic(!) and I agree that looks like a really good implementation, excellent use of space and very little extra bloat to the view. :sunglasses: I think delivering at least some of the info in a zero click manner like that then clicking the 'more' (expanding the package card?) would be preferable to having to click first before seeing anything.

As a bonus that kind of up front display may induce some of those packages that don't currently employ changelogs to do so. :)

**Comment by benogle** [comment]
Would this be automated or rely on markdown in the body of the release?

**Comment by mnquintana** [comment]
I'm thinking we should have a series of fallbacks - first check for release notes via GitHub Releases (which would be the easiest, since we wouldn't have to parse anything), then check for a `CHANGELOG.md`, then a `CHANGELOG.txt` or just `CHANGELOG`. For the files, we could either make some attempt to parse them and only display changes for that version (of course, this relies on the changelogs being well-formed and using headers semantically), or just don't bother and display a truncated version of the entire file.

**Comment by 50Wliu** [comment]
A last-ditch fallback could be to just list the raw commit messages.

**Comment by simurai** [comment]
Hmm.. I'd rather not show anything, than adding potentially too much noise. As a convention it could be this (if that's possible):

1. Parse `CHANGELOG.md`.
2. Look for the version that the update is for.
3. Show all text till the currently installed version.

As an example with linter. Let's say I'm still using `1.7.2`. The latest update would be for `1.8.1`. So I would see all this text:

```
## 1.8.1

* Workaround an atom bug where moving an editor between panes would throw an error
* Allow panel resize to persist
* Add option to always have the panel to fill the minimum space
* Fix issue where external scope errors would cause panel to continue displaying

## 1.8.0

* Improve rendering of multiline messages to align with recent single line changes
* Enable clicking on multiline messages to view next lines
* Add default styles for `Info` type
* Aligned bottom panel buttons
* Made the text in the bottom panel selectable
* Linter bottom panel is now resizable (Drag the height down to 0 to reset your changes)
* Add a new `class` attribute for providers (when class is provided, no class is automatically added).
* Fix linter gutter dot alignment across themes and zoom level.
```

No formatting, just plain text. Truncated after 3-4 lines.

**Comment by simurai** [comment]
> first check for release notes via GitHub Releases (which would be the easiest, since we wouldn't have to parse anything)

That would be nice too. Not sure if there should be a push to use either GitHub's release notes or `CHANGELOG.md` so it would be more consistent.

**Comment by thedaniel** [comment]
I think we should choose a single blessed source and promote it, not try to handle every case. GitHub Releases is appealing--we could even revive the "publish to atom.io via Release hook" feature. I'm not sure how many people use Releases as heavily as we do, or even want to though.

Parsing the changelog is nice because it's a file in the repo, but Releases are tied to tags and there's no chance of people messing up the syntax and having changelogs we can't figure out how to parse.

**Comment by lee-dohm** [comment]
We already support `CHANGELOG.md` even though it is just a link on the package's Settings View page:

<img width="622" alt="screen shot 2015-10-07 at 3 38 55 pm" src="https://cloud.githubusercontent.com/assets/1038121/10353230/a30a0be4-6d09-11e5-990c-0fb9d0aa212b.png">

So it seems a bit lame to have package authors go back and transcribe all of that information into Releases. If we are going to go the `CHANGELOG.md` route, we should assume it is Markdown and just convert it and display it. We don't have to mandate a format or anything, just show some HTML and let the package author handle how they want to format their CHANGELOG.

On the flip side, I like Releases and think it would be easier to support in the long run. So maybe it is a bit of short-term pain for a better solution in the long run.

**Comment by danielbrodin** [comment]
Using `apm publish` is really convenient, but is there a way to add a description to the tag through it, or would using GitHub Releases mean we would have to manually do a release, or do `apm publish` then go to GitHub and edit the release description?
Haven't used GitHub Releases except for the once created through `apm publish`, so just asking.

Parsing the `CHANGELOG.md` could probably be a bit tough since there are so many different formattings, and you would probably only want to show what have changed since you last did an update/install. If the data should be in a file in the repo instead of GitHub Releases maybe a `CHANGELOG.json` or something would be better?

**Comment by zenopopovici** [comment]
Couldn't we support both? If one is empty then show the other? This way we don't have to change anybody's work style, and we wouldn't make a decision that's for the developer to make. I don't think it would complicate things too much. Just a thought ...

**Comment by thedaniel** [comment]
In my experience, in general, it's best to start with one workflow and expand from there in the future rather than trying to support multiple workflows from the start.

**Comment by thomasjo** [comment]
> I think we should choose a single blessed source and promote it

What about [Keep a Changelog](http://keepachangelog.com/):question:

**Comment by laughedelic** [comment]
+1 for the changes list preview on update and for the changelog button next to update button

**Comment by dorian-marchal** [comment]
:+1: to show the changelog directly on the "update package" card.

**Comment by mediabeastnz** [comment]
:+1: List of changes will be so handy!

**Comment by iamstarkov** [comment]
it will be so cool, if this will be implemented

**Comment by ai** [comment]
I think Changlog is very important for every project and every update, user should do.

**Comment by burodepeper** [comment]
This issue has been silent for quite a while, but I'd like to weigh in with a few cents of my own.

As a package-author I'm looking for a way to communicate simple messages (about new functionality or serious bugs, for instance) to users. My first instinct was not to tie this to the package-update-card (but rather to opening a particular type of document after updating), but now I'm thinking _why not?_.

As a user I'm confronted with more updates than I'd like, and I'm a little reluctant to click "update all packages" because of mishaps in the past. (on a side-note, I wish I could skip updates)

I think CHANGELOGs are great, but as a package-author I'd prefer more control of the message. Also, my intended user-persona is not the same as who I imagine inspects CHANGELOGs. Besides all of that, incremental updates are an issue. I agree with @simurai that it's best to show as little as possible, than it is to create unnecessary "noise".

I like the idea of using `apm publish <major|minor|patch> <message>`, and I think visually distinguishing between major, minor and patch updates could help in some way.

---

If I were dictator: enforce properly formatted CHANGELOGs (if package-authors wish to use this feature anyway), and determine the amount of content to the `weight` of the update.

**Comment by jrie** [comment]
I also would like to bump this enhancement.

It would be great to have a such information shown on the update package page, controlled by a button to close/open details. Similar to what Firefox's does for example with the Add-On update notes which can be expanded and or collapsed on demand.

**Comment by IanCaunce** [comment]
The main issue for me is that you can't view the changelogs before the package has been updated. There is also no recently update list, meaning that if you click update all, you have to remember what packages have been updated and view their changelogs post update. Alternatively you can go to the github repo or the atom.io page for the package and view the changes there prior to updating.

There is no easy and convenient way of seeing what new features are being brought in before updating. 

Recently there was an update to https://github.com/atom-minimap/minimap which added ["Kite promotion"](https://github.com/atom-minimap/minimap/commit/16c11d82b889ce1260342e4fa7d6d1905c0fde45). If I had seen this before to updating, I wouldn't have updated the package immediately. As I saw this after updating the package, I had a small heart attack as to me it sounded like some kind of adware/grayware had been bolted onto the package which I had already installed. The promotion was declared inside the CHANGELOG.md file for the package so I could have been avoided it if I was able to view the changelogs before updating.
